ops/c/src/core/ops_dummy_singlenode.cpp

```diff
@@ -626,88 +551,97 @@ void ops_dat_fetch_data_host(ops_dat dat, int part, char *data) {
-  if (dat->block->instance->OPS_soa && dat->dim > 1) throw OPSException(OPS_NOT_IMPLEMENTED, "Error, missing OPS implementation: ops_dat_fetch_data not implemented for SoA");
-  copy_loop<0>(data, dat->data, lsize, dat->size, dat->d_m, dat->elem_size);
+  if (dat->block->dims > 5)
+    throw OPSException(OPS_NOT_IMPLEMENTED,
+                       "Error, missing OPS implementation: ops_dat_fetch_data "
+                       "not implemented for dims>5");
+  fetch_loop_slab(data, dat->data, lsize, dat->size, dat->d_m, dat->elem_size,
+                  dat->dim, range2);
 }
 
-
-void ops_dat_fetch_data_slab_host(ops_dat dat, int part, char *data, int *range) {
-    (void)part;
+void ops_dat_fetch_data_slab_host(ops_dat dat, int part, char *data,
+                                  int *range) {
+  (void)part;
   ops_execute(dat->block->instance);
   ops_get_data(dat);
   int lsize[OPS_MAX_DIM] = {1};
-  int range2[2*OPS_MAX_DIM] = {0};
+  int range2[2 * OPS_MAX_DIM] = {0};
   for (int d = 0; d < dat->block->dims; d++) {
-    lsize[d] = range[2*d+1]-range[2*d+0];
-    range2[2*d] = range[2*d];
-    range2[2*d+1] = range[2*d+1];
-
+    lsize[d] = range[2 * d + 1] - range[2 * d + 0];
+    range2[2 * d] = range[2 * d];
+    range2[2 * d + 1] = range[2 * d + 1];
   }
   for (int d = dat->block->dims; d < OPS_MAX_DIM; d++) {
     lsize[d] = 1;
-    range2[2*d] = 0;
-    range2[2*d+1] = 1;
+    range2[2 * d] = 0;
+    range2[2 * d + 1] = 1;
   }
-  lsize[0] *= dat->elem_size; //now in bytes
-  if (dat->block->dims>5) throw OPSException(OPS_NOT_IMPLEMENTED, "Error, missing OPS implementation: ops_dat_fetch_data not implemented for dims>5");
-  if (dat->block->instance->OPS_soa && dat->dim > 1) throw OPSException(OPS_NOT_IMPLEMENTED, "Error, missing OPS implementation: ops_dat_fetch_data not implemented for SoA");
-  copy_loop_slab<0>(data, dat->data, lsize, dat->size, dat->d_m, dat->elem_size, range2);
+  if (dat->block->dims > 5)
+    throw OPSException(
+        OPS_NOT_IMPLEMENTED,
+        "Error, missing OPS implementation: ops_dat_fetch_data_slab_host "
+        "not implemented for dims>5");
+  fetch_loop_slab(data, dat->data, lsize, dat->size, dat->d_m, dat->elem_size,
+                  dat->dim, range2);
 }
 
-
 void ops_dat_set_data(ops_dat dat, int part, char *data) {
   ops_dat_set_data_memspace(dat, part, data, OPS_HOST);
 }
 
-
 void ops_dat_set_data_host(ops_dat dat, int part, char *data) {
   ops_execute(dat->block->instance);
-  int lsize[OPS_MAX_DIM] = {1};
-  int ldisp[OPS_MAX_DIM] = {0};
-  ops_dat_get_extents(dat, part, ldisp, lsize);
-  for (int d = dat->block->dims; d < OPS_MAX_DIM; d++) {
-    lsize[d] = 1;
-    ldisp[d] = 0;
+
+  int *range{new int(2 * dat->block->dims)};
+  for (int d = 0; d < dat->block->dims; d++) {
+    range[2 * d] = dat->d_m[d];
+    range[2 * d + 1] = dat->size[d] - dat->d_p[d];
```

Comment:
I am not sure this range calculation is right - dat->size[d] is compute as user_given_size[d]-d_m[d]+d_p[d] what is the intent?

Reply:
in that case, the second line should be dat->size[d] + dat->d_m[d]; Please make sure this is consistent with the MPI version - but there the intra-block MPI halos are of course not included